crmsh/corosync.py

```diff
@@ -512,3 +514,309 @@ def remove_key(cls, path, index=0):
+
+        Returns:
+            a reference to in/out arg `config`
+        """
+        existing_addr_node_map = {
+            utils.IP(node.addr).ip_address: node.nodeid
+            for link in links if link is not None
+                for node in link.nodes
+            if node.addr != ''
+        }
+        for nodeid, addr in node_addresses.items():
+            found = next((node for node in links[linknumber].nodes if node.nodeid == nodeid), None)
+            if found is None:
+                raise ValueError(f'Unknown nodeid {nodeid}.')
+            canonical_addr = utils.IP(addr).ip_address
+            if (
+                    found.addr == ''    # adding a new addr
+                    or utils.IP(found.addr).ip_address != canonical_addr    # updating a addr and the new value is not the same as the old value
+            ):
+                # need to change uniqueness
+                existing = existing_addr_node_map.get(canonical_addr, None)
+                if existing is not None:
+                    raise LinkManager.DuplicatedNodeAddressException(addr, nodeid, existing)
+            found.addr = addr
+            existing_addr_node_map[canonical_addr] = found.nodeid
+        nodes = config['nodelist']['node']
+        assert isinstance(nodes, list)
+        for node in nodes:
+            updated_addr = node_addresses.get(int(node['nodeid']), None)
+            if updated_addr is not None:
+                node[f'ring{linknumber}_addr'] = updated_addr
+        return config
+
+    def add_link(self, node_addresses: typing.Mapping[int, str], options: dict[str, str|None]) -> dict:
+        links = self.links()
+        next_linknumber = next((i for i, link in enumerate(links) if link is None), -1)
+        if next_linknumber == -1:
+            raise ValueError(f'Cannot add a new link. The maximum number of links supported is {KNET_LINK_NUM_LIMIT}.')
+        nodes = next(x for x in links if x is not None).nodes
+        unspecified_nodes = [node.nodeid for node in nodes if node.nodeid not in node_addresses]
+        if unspecified_nodes:
+            raise self.MissingNodesException(unspecified_nodes)
+        links[next_linknumber] = Link(next_linknumber, [dataclasses.replace(node, addr='') for node in nodes])
+        self.__upsert_node_addr_impl(self._config, links, next_linknumber, node_addresses)
+        return self.update_link(next_linknumber, options)
+
+    def remove_link(self, linknumber: int) -> dict:
+        """Remove the specified link.
+
+        Parameters:
+            * linknumber: the link to update
+        Returns: updated configuration dom. The internal state of LinkManager is also updated.
+        """
+        links = self.links()
+        if links[linknumber] is None:
+            raise ValueError(f'Link {linknumber} does not exist.')
+        if sum(1 if link is not None else 0 for link in links) <= 1:
+            raise ValueError(f'Cannot remove the last link.')
```

Comment:
```suggestion
            raise ValueError(f'Cannot remove the last link.')

        _, outp = ShellUtils().get_stdout("corosync-cfgtool -n")
        logger.debug(f"remove_link( {linknumber} ) output = {outp}")

        is_linknumber_connected = len(re.findall(f'LINK: {linknumber} .* connected', outp)) == 1
        is_only_one_link_connected = len(re.findall(f'LINK: .* connected', outp)) == 1

        if is_linknumber_connected and is_only_one_link_connected:
            raise ValueError(f'Cannot remove the last "connected" link.')

```

Reply:
This patch works only in a 2-node cluster. With multi-link X multi-node in knet, it is too difficult for crmsh to determine which link should not be removed at runtime.

Reply:
You are right. This patch suggestion doesn't fly with 3+ node cluster. 

One step back, the point is "remove" operation need more care, need avoid intrusive operation for the running cluster. Let's think about it...

Reply:
Here is my update. 

For all endpoints in the same link, one "disconnected" will populate to all endpoints.
That says the state of all endpoints in the same knet link is either all "connected" or all "disconnected" in the same node partition. 

With that, the code can be quite neat to work with.

Reply:
> For all endpoints in the same link, one "disconnected" will populate to all endpoints.

No, this is incorrect. Knet links are not totem rings. When there are multiple links, a node can use different links to communicate with different nodes. For example, the following setup forms a 3 node quorum with each link having disconnected endpoints:

```
suse@ha-3-1:~> sudo corosync-cfgtool -n
Local node ID 1, transport knet
nodeid: 2 reachable
   LINK: 0 udp (192.168.122.73->192.168.122.74) enabled connected mtu: 1397
   LINK: 1 udp (192.168.123.73->192.168.123.74) enabled mtu: 1397
   LINK: 2 udp (192.168.124.73->192.168.124.74) enabled mtu: 1397

nodeid: 3 reachable
   LINK: 0 udp (192.168.122.73->192.168.122.75) enabled mtu: 1397
   LINK: 1 udp (192.168.123.73->192.168.123.75) enabled mtu: 1397
   LINK: 2 udp (192.168.124.73->192.168.124.75) enabled connected mtu: 1397

suse@ha-3-1:~> sudo corosync-quorumtool
Quorum information
------------------
Date:             Mon Aug  5 10:48:23 2024
Quorum provider:  corosync_votequorum
Nodes:            3
Node ID:          1
Ring ID:          1.d
Quorate:          Yes

Votequorum information
----------------------
Expected votes:   3
Highest expected: 3
Total votes:      3
Quorum:           2  
Flags:            Quorate 

Membership information
----------------------
    Nodeid      Votes Name
         1          1 ha-3-1 (local)
         2          1 ha-3-2
         3          1 ha-3-3
```

Reply:
Implementing validation on this provides little values to users. Link management is mostly used for setting up a cluster, and is not a daily task for operating a cluster. And pcs does not validate this either.

And this is not only a problem of a bit more lines of code. `corosync-cfgtool` gives only the information the connections between local node and other nodes. The connection status between other nodes is missing. Even we get all the information needed, to predict whether certain links can form a quorum, we needed to implement the same algorithm used in totem and knet. And if the implementation had any difference in detail with the one used in corosync, it would give a wrong answer and this would be worse than not validating the links.

Reply:
I agree and see the gap it is does not resolve the ideal situation as you pointed out.

I give it another try, the code is could be simple enough for the this typical situation to prevent error.